src/Engine/ProtoCore/Lang/FunctionEndPoint.cs

```diff
@@ -7,6 +8,231 @@
+                //if (!FunctionGroup.CheckInvalidArrayCoersion(fep, reducedParamSVs, allowArrayPromotion))
+                return dist;
+            }
+
+            return (int)ProcedureDistance.InvalidDistance;
+        }
+
+        internal int ComputeCastDistance(List<CLRStackValue> args)
+        {
+            //Compute the cost to migrate a class calls argument types to the coresponding base types
+            //This cannot be used to determine whether a function can be called as it will ignore anything that doesn't
+            //it should only be used to determine which class is closer
+            if (args.Count != FormalParams.Count)
+            {
+                return int.MaxValue;
+            }
+
+            if (0 == args.Count)
+            {
+                return 0;
+            }
+
+            int distance = 0;
+            for (int i = 0; i < args.Count; ++i)
+            {
+                int rcvdTypeId = args[i].TypeUID;
+                System.Type rcvdType = args[i].Type;
+
+                // If its a default argumnet, then it wasnt provided by the caller
+                // The rcvdType is the type of the argument signature
+                if (args[i].IsDefaultArgument)
+                {
+                    rcvdTypeId = FormalParams[i].UID;
+                    rcvdType = FormalParams[i].CLRType;
+                }
+
+                var expectedType = FormalParams[i];
+                if (expectedType.IsIndexable != args[i].IsEnumerable) //Replication code will take care of this
+                {
+                    continue;
+                }
+                else if (expectedType.IsIndexable)  // both are arrays
+                {
+                    continue;
+                }
+                else if (expectedType.UID == rcvdTypeId)
+                {
+                    continue;
+                }
+                else if (rcvdTypeId != Constants.kInvalidIndex && expectedType.UID != Constants.kInvalidIndex)
+                {
+                    int currentCost = ClassUtils.GetUpcastCountTo(
+                        (rcvdTypeId, rcvdType),
+                        (expectedType.UID, expectedType.CLRType));
+                    distance += currentCost;
+                }
+            }
+            return distance;
+        }
+
+        internal int ComputeTypeDistance(List<CLRStackValue> args, MSILRuntimeCore runtimeCore, bool allowArrayPromotion = false)
+        {
+            if (args.Count == 0 && FormalParams.Count == 0)
+            {
+                return (int)ProcedureDistance.ExactMatchDistance;
+            }
+
+            if (args.Count != FormalParams.Count)
+            {
+                return (int)ProcedureDistance.MaxDistance;
+            }
+
+            int distance = (int)ProcedureDistance.MaxDistance;
+            // Jun Comment:
+            // Default args not provided by the caller would have been pushed by the call instruction as optype = DefaultArs
```

Comment:
does this comment still apply?

Reply:
I did not try to adapt the old functions to the new VM except for the bare minimum.
We can adjust comments and redundancies later on